### ISSUE TITLE
I lost Y axis completely after driver install

### ISSUE THREAD

**Issue by cer0-0** [AndyFilter/YeetMouse#13]
Hi;

For some reason, when I install via install.sh, I completely lost Y axis input for Razer Viper Ultimate mice. It works as intended for Logitech G502 and it doesn't recognize at all when I plug in logitech G305. I have no idea what might be wrong with it or how am I supposed to fix it, only thing that comes to my mind is razer has seperate X and Y dpi sliders while logitech has a unified dpi but I'm not sure if it matters. For g305 i have literally no idea.

Not really related to the problem but is it possible to have angles go up to 360 instead of 180 since i need something between -1.7 to -2 degrees(should be the same to 358.x degrees) to fix my mouse posture problem. I don't think I'm the only one who needs negative rotation in the world.

Anyways, thanks for doing something many just don't bother with. I'm literally just starting my linux journey and its kind of a bumpy ride, but things like these projects helps coping a lot. If there is anything I could help I would love to try.

Best regards

**Comment by WriteMeaCadenza** [comment]
i would also like to report i have this happen with certain mice as well, let me know if theres anything i can do or provide to help you address the issue when you get around to it

thanks for making a comparable program to raw accel

**Comment by pxlsec** [comment]
I have the same problem on my mouse `pulsar x2 wireless`.

I took the time to bisect and found that ba430f6da3f53b318505590e732cbf0c6031d315 was the first bad commit.

**Comment by AndyFilter** [comment]
Hey, Thanks for sharing your experience and sorry that I took so long to respond, health issues.
First of all, I'm glad you found this program useful and that it helps you get used to Linux. 

About the issue, it's not the first time I see something like this happening, there are known issues with more "_fancy_" mice, so stuff like colorful LED's, a bunch of buttons, etc. this can cause issues because of the current way the HID descriptors are parsed (it's partially fixed, but as you can see, not completely). The problem with this _problem_ is that I can't reproduce it on my end and find the underlying issue, as I don't have a mouse that would cause it (trust me I tried with quite a few mice).

To start off I'd like everyone involved and willing to help send me some terminal screenshots which you can get by:
1. (re-)installing the driver.
2. running `sudo dmesg` to show the driver logs (You might need to scroll up a bit).

At the end You should have something like this:
![image](https://github.com/user-attachments/assets/9fccfe0d-875e-44bb-bf83-b17fbc4d22ea)

If You have more than one mouse plugged in, make sure you send all of them, or at least the one that's **not** working.


When it comes to the rotation, first of all, thanks for letting me know; secondly, I can set the range to [-180, 180] (rather than go up to 360), let me know if that would fix the issue. Remember that you can manually edit all value "sliders" by clicking them while holding _CTRL_. Let me know if the negative values work as expected and I'll apply the changes.

**Comment by AndyFilter** [comment]
> I have the same problem on my mouse `pulsar x2 wireless`.
> 
> I took the time to bisect and found that [ba430f6](https://github.com/AndyFilter/YeetMouse/commit/ba430f6da3f53b318505590e732cbf0c6031d315) was the first bad commit.

Hey @pxlsec, thanks for pinning down the exact commit, I'm sure that this commit doesn't work as there is a pretty big issue (unrelated to this one) with it (please uninstall the driver and reboot immediately if you've installed this commit), but the next commit fixed the most obvious issue with this one. So please don't use this commit, and pretend like it doesn't exist haha. At the end I'm pretty sure it's all about the "NUM_USAGES" value here:
https://github.com/AndyFilter/YeetMouse/blob/df1e44b18522c483d8858b85dd5887798a7d2867/driver/util.c#L24
If You set it to a big enough value (something like 128, or even 256) everything ***should*** work fine. But that's not really a solution isn't it? The _real_ solution (or at least so I hope) is being worked on here #5, and some great efforts are being put there to make sure we can have a fast and reliable driver. So once that's done, this issue will hopefully be solved as well.

**Comment by pxlsec** [comment]
> (please uninstall the driver and reboot immediately if you've installed this commit), but the next commit fixed the most obvious issue with this one. So please don't use this commit

You see i think I'm having the same issue as the author (not being able to move the cursor up/down). As for the commit, this was simply the commit that broke y axis (at least for me). And the problems still persists on master. I tested on both `6.12.4-linux-zen` and `6.12.1-273-linux-tkg` (tkg compiled using clang).
The result of `dmesg` (on df1e44b18522c483d8858b85dd5887798a7d2867 aka latest master):
```
[30106.013654] usbcore: registered new interface driver leetmouse
[30106.099956] LEETMOUSE: New r_count (5) is equal or smaller than the max r_count (64)
[30106.099958] LEETMOUSE: New r_count (1) is equal or smaller than the max r_count (64)
[30106.099959] LEETMOUSE: New r_count (2) is equal or smaller than the max r_count (64)
[30106.099960] LEETMOUSE: New r_count (1) is equal or smaller than the max r_count (64)
[30106.099960] LEETMOUSE: New r_count (1) is equal or smaller than the max r_count (64)
[30106.099985] input: Compx X2 Wireless as /devices/pci0000:00/0000:00:14.0/usb1/1-8/1-8:1.0/input/input28
[30106.113714] EDAC skx: ECC is disabled on imc 0
```

**Comment by AndyFilter** [comment]
Thanks for the console printout, it helped a lot. I have created a [new branch](https://github.com/AndyFilter/YeetMouse/tree/driver/experimental) for problems like this one, and linked it to this issue. And by looking at your logs I've managed to track down some possible issues with the code, that are now fixed on the new branch. You might wanna check them out, and let me know if these changes did anything. Please keep in mind tho, that this is an experimental branch, and even tho I only changed 2 lines, something else might have been affected by this, so be on a lookout for other unexpected behavior. Everything works flawlessly for me, but I had no issue earlier either.

Btw. the issue (at least one that I fixed now) was with increasing sizes of the _r_count_ in the hid parser. I assumed that the size values would only be getting smaller, turns out I was wrong.
> [30106.099958] LEETMOUSE: New r_count (**1**) is equal or smaller than the max r_count (64)
> [30106.099959] LEETMOUSE: New r_count (**2**) is equal or smaller than the max r_count (64)

As you can see the "New r_count" got bigger.

The only thing I don't understand is why it always caused the Y axis to stop working...

**Comment by pxlsec** [comment]
The fix on experimental does indeed do the trick (I think @cer0-0 should try aswell). As for other issues, mouse acceleration works fine after installing and I see no obvious errors in dmesg or journal.
I am however getting a GLFW error trying to open the gui now (both on master and experimental) so it's unrelated to this issue and probably just something I did.
``` sh
❯ sudo ./gui/YeetMouseGui
error: XDG_RUNTIME_DIR is invalid or not set in the environment.
Authorization required, but no authorization protocol specified

GLFW Error 65550: Failed to detect any supported platform
YeetMouseGui: External/ImGui/imgui.cpp:4432: ImGuiIO& ImGui::GetIO(): Assertion `GImGui != __null && "No current context. Did you call ImGui::CreateContext() and ImGui::SetCurrentContext() ?"' failed.
zsh: IOT instruction  sudo ./gui/YeetMouseGui
```

**Comment by WriteMeaCadenza** [comment]
Hi, Sorry for the late reply i was asleep. i've taken a couple of screenshots because i still had the driver installed and i'm not sure if that would affect the information you're looking for but i noticed the information displayed a little different. so i might as well send it all incase. i have a gamebitions orbit mouse

before i reinstalled the driver from reboot;
![Screenshot from 2024-12-22 13-04-34](https://github.com/user-attachments/assets/40abb46f-1831-47cd-b1ea-857ac0d37813)
after replugging the mouse back in;
![Screenshot from 2024-12-22 13-07-53](https://github.com/user-attachments/assets/0b970c86-1700-4b49-a2e7-623c22395cd7)
1a37e)
after driver reinstall; 
![Screenshot from 2024-12-22 13-21-19](https://github.com/user-attachments/assets/0062f59d-1f62-44c5-aed2-3dd1fceeb96b)

**Comment by AndyFilter** [comment]
@pxlsec Thanks for trying the _experimental_ version, I'm glad it worked! I'll merge the branches once I'm sure it's working correctly for more people.
As for the GUI issues try to run it as stated in the instructions, which is `sudo -E ./YeetMouseGui`, or in Your case: `sudo -E ./gui/YeetMouseGui`. The `-E` flag simply preserves the environment variables in *sudo*.

@WriteMeaCadenza Thanks for the logs, this confirms pretty much everything. The growing "r_count" was causing the issue. You can now test the [new branch](https://github.com/AndyFilter/YeetMouse/tree/driver/experimental) I've created for this issue that should fix it here: https://github.com/AndyFilter/YeetMouse/tree/driver/experimental. Let me know if it helped, and if not, please share the logs with me.

**Comment by WriteMeaCadenza** [comment]
Yep, that fixed it all up. thank you!

**Comment by WriteMeaCadenza** [comment]
actually sorry i may have spoke too soon. upon restart it's not working again. 
![Screenshot from 2024-12-25 23-05-12](https://github.com/user-attachments/assets/8a673a03-5223-4fdf-828f-09c4a29b0c7d)
![Screenshot from 2024-12-25 23-03-55](https://github.com/user-attachments/assets/8b5c501b-dfe5-488f-8394-c36e9defec69)

**Comment by AndyFilter** [comment]
@WriteMeaCadenza Interesting, the logs don't seem to be different, or at least the part I'm interested in. So, just to clear things out, You uninstalled the old version of the driver, installed the new version of the driver (it was working fine at this moment), rebooted the PC and the Y axis stopped working? (also, are you using the `install.sh` / `uninstall.sh` scripts?)

Can you try reinstalling the driver again from the [driver/experimental](https://github.com/AndyFilter/YeetMouse/tree/driver/experimental) branch? I want to pin down where the issue is coming from, the reboot, or the driver itself.

Does reconnecting the mouse help?

**Comment by WriteMeaCadenza** [comment]
yeah it is a bit weird and that's correct i uninstalled the old driver via the uninstall.sh script removed the old folder replaced it with the experimental version and installed it via the install.sh script it was working as expected i think i had even restarted it on the day and it was fine went to sleep and when i tried it the next day the y axis wasn't working again.

also can confirm when i reinstall the experimental mouse driver again it starts working as expected. The most odd thing is now i can't get it to replicate it and it seems to be okay on restart i am still getting the "shift out of bounds. shift exponent -1 is negative" error show up in my logs though

i am new to linux and trying to sort through some issues i am having with other things on my operating system maybe i accidentally did something. do you want me to send any other logs if it starts acting up again?

**Comment by AndyFilter** [comment]
Alright, so it's working _in the end_. Don't worry if you're new to Linux and you don't understand some things, I was there (and still am, to be honest), that's exactly the reason why I wanted to make this program, to help people that are just starting their journey with Linux.

Also, don't worry about the "Shift out of bounds..." _error_, it's just a one time thing (happens every restart / reinstall).

If anything unexpected / unwanted happens, then sure, send the logs and inform me about it.

**Comment by AndyFilter** [comment]
@WriteMeaCadenza Since You brought up the topic of the "Shift out of bounds" here, I fixed it and pushed a quick (2 lines) patch (5bdabfd) to the [driver/experimental](https://github.com/AndyFilter/YeetMouse/tree/driver/experimental) branch.
I know the issue is unrelated, but I don't want to open a new ticket for something this simple, so if anyone can try this out and let me know if it's working as intended without any side effects, I'll be grateful.

**Comment by WriteMeaCadenza** [comment]
fair enough, just gave it an install and i'm not seeing it pop up in my logs. appreciate all your work!

**Comment by AndyFilter** [comment]
I would like to close this issue as it seems to be resolved. Can You please confirm that everything is working as expected, @cer0-0?
